### PR TITLE
Add arbitrary config tests for function DDL statements

### PR DESCRIPTION
**Pull request by hanefi**

### CODE REVIEW

**Comment by hanefi** [comment]
I had several issues during writing these tests:

1. Arbitrary config tests are not run with a superuser unlike the rest of our regression suite. `LEAKPROOF` option can only be set by superuser, and I failed to do that when running tests on my local setup. CI environment however uses superuser, and creates a different output.
2. Arbitrary config tests are affected by `~/.psqlrc` configs unlike the rest of our regression tests. Hence I get many error messages such as
`psql:/home/hanefi/.psqlrc:3: ERROR:  role "hanefi" does not exist`
3. Arbitrary config base schedule and regular schedule has different verbosity. Hence the need to use `terse` option
4. We do not normalize the ports and other values, and this makes it harder to write tests that generate the same output.

**Comment by onderkalaci** [comment]
Thanks for the detailed analysis: 


> Arbitrary config tests are not run with a superuser unlike the rest of our regression suite. LEAKPROOF option can only be set by superuser, and I failed to do that when running tests on my local setup. 

Not running as a superuser is a design choice we made. Because, in production workloads superusers are rarely used, and arbitrary config made a lot of sense not to run as superuser.

> CI environment however uses superuser, and creates a different output.

That sounds unexpected. We only have one config with superuser to have coverage: https://github.com/citusdata/citus/blob/b092a1a496dfd752543e10e0e481fcdbb48d6fe2/src/test/regress/citus_tests/config.py#L207-L210  

Did you maybe see this one failing? Or all of them? If the latter, we should definitely look what is going on, it might be  bug in the arbitrary configs.

> Arbitrary config tests are affected by ~/.psqlrc configs unlike the rest of our regression tests. Hence I get many error messages such as
psql:/home/hanefi/.psqlrc:3: ERROR: role "hanefi" does not exist

Can you open an issue to track this? Not sure why that is different than regular regression tests.

> Arbitrary config base schedule and regular schedule has different verbosity. Hence the need to use terse option

Can we add the `terse` option to a more common place if possible (e.g., as a pg config setting to base class)? We should aim for minimal verbosity at any arbitrary config schedule. 

> We do not normalize the ports and other values, and this makes it harder to write tests that generate the same output. This is especially important when delegating function calls that create flaky outputs.

I think that should never need to normalize the outputs. The idea is that we should be indistinguishable from regular PG in this tests. Why would function delegation produce extra output if logs are not enabled?

File: src/test/regress/create_schedule [hunk 1]
```diff
@@ -8,6 +8,7 @@ test: schemas_create
 test: views_create
 test: sequences_create
 test: index_create
+test: function_create
```
Comment (onderkalaci) [inline src/test/regress/create_schedule:11]: I think the `create` piece is almost ready to go. But, one of the main goals of arbitrary configs is to also add a tests to non-create schedule such that they are executed from the worker nodes with all different kinds of configs.

For example, `SELECT warning(message) FROM warnings WHERE id = 1;` should go there and potentially few more executions. Like, a function does bunch of commands. 

I'd even consider getting HammerDB functions to this test, and have two copies each. The first copy is a distributed function/procedure, the other is non-distributed procedure. Both should produce the same output (and on pg as well).

If we do hammerDB, we might need to add `create_distributed_function` here as a no-op of Postgres schedule: https://github.com/citusdata/citus/blob/master/src/test/regress/sql/postgres.sql

**Comment by hanefi** [comment]
> Not running as a superuser is a design choice we made. Because, in production workloads superusers are rarely used, and arbitrary config made a lot of sense not to run as superuser.

Thanks for the explanation. I did not know that this was intentional. I see how this decision can be valuable now.

> > CI environment however uses superuser, and creates a different output.
> 
> That sounds unexpected. 

`LEAKPROOF` option can only be set by superusers. I created a test query, and committed the local test output. In [CI test run](https://app.circleci.com/pipelines/github/citusdata/citus/23487/workflows/ce55f07a-89f2-4df3-ad62-5ac3918ab2ba/jobs/640571) , I got the following diff:

```
 ALTER FUNCTION eq(macaddr,macaddr) CALLED ON NULL INPUT IMMUTABLE SECURITY INVOKER PARALLEL UNSAFE LEAKPROOF COST 5;
-ERROR:  only superuser can define a leakproof function
```

Hence, I assumed that CI is actually running the tests with superuser connection. I ended up removing tests with `leakproof` clause as I could not get the Ci and local setup behave the same here.

> Why would function delegation produce extra output if logs are not enabled?

Valid question, and good advice. I guess I should be fine if I decrease log levels, and verbosity.

File: src/test/regress/citus_tests/config.py [hunk 2]
```diff
@@ -321,11 +321,14 @@ def __init__(self, arguments):
...
+            # when citus.shard_replication_factor >= 2
             "arbitrary_configs_truncate_partition_create", "arbitrary_configs_truncate_partition",
             # citus does not support modifying a partition when
-            # citus.shard_replication_factor > 2
-            "arbitrary_configs_truncate_cascade_create", "arbitrary_configs_truncate_cascade"]
+            # citus.shard_replication_factor >= 2
+            "arbitrary_configs_truncate_cascade_create", "arbitrary_configs_truncate_cascade",
+            # citus does not support colocating functions with distributed tables when
+            # citus.shard_replication_factor >= 2
+            "function_create", "functions"]
```
Comment (hanefi) [inline src/test/regress/citus_tests/config.py:331]: I could consider splitting these files into two. Not all the tests there require we have `shard_replication_factor=1`

Comment (hanefi) [reply src/test/regress/citus_tests/config.py:331]: It fails when we supply the distribution column argument.


```diff
 SELECT create_distributed_function('warning(int,text)','$1');
- create_distributed_function
----------------------------------------------------------------------
-
-(1 row)
-
+ERROR:  cannot colocate function "warning" and table "table_1"
```

The error message comes from https://github.com/citusdata/citus/blob/dd78c813781cd4ebc32bb049935a1a8f104693c4/src/backend/distributed/commands/function.c#L656-L668

File: src/test/regress/expected/function_create.out [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    RETURNS bool
+    LANGUAGE plpgsql
+AS $func$
+DECLARE
+    coordinatorSql text;
+    workerSql text;
+BEGIN
+    SELECT pg_get_functiondef(funcname::regprocedure) INTO coordinatorSql;
+    FOR workerSql IN SELECT result FROM run_command_on_workers('SELECT pg_get_functiondef(' || quote_literal(funcname) || '::regprocedure)') LOOP
+            IF workerSql != coordinatorSql THEN
```
[the remark this answers is not in the document]
Comment (hanefi) [reply src/test/regress/expected/function_create.out:15]: It is basically no-op as we compare the function definition with itself. I guess that is enough to see that an `ALTER FUNCTION` is successful. No need to compare anything on single node.

File: src/test/regress/citus_tests/config.py (hunk 2 above)
Comment (onderkalaci) [reply src/test/regress/citus_tests/config.py:331]: I think we can keep like this, no need to split. Thanks

File: src/test/regress/expected/function_create.out (hunk 3 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/test/regress/expected/function_create.out:15]: Ah, okay we iterate on the output of `run_command_on_workers`. Given that `run_command_on_workers` returns no rows, we never find non-equal

**Review by onderkalaci — APPROVED** [review]

**Pull request merged**